File: sdk/js/src/groups.ts
// Group helpers (JS-007).
//
// Mirrors `sdk/python/pageros/groups.py`. Multi-device apps emit group
// events (member_joined/left, presence_update, group_message — SPEC
// §5.4.2). Online subscribers consume them via the app's own long-poll
// endpoint; offline subscribers receive them through the Push Relay's
// `/group_push` endpoint with a per-recipient X25519+ChaCha20 envelope.
//
// What this module owns:
//
//   - The wire path + content type for the group-push batch
//     (`POST /group_push`, application/json).
//   - The recipient-result enum so callers can branch on
//     "accepted" vs "rate_limited" vs "bad payload" etc.
//   - `sendGroupPush` — encrypt one payload per recipient, sign the
//     whole batch, POST. Returns the per-recipient result.
//
// What this module does NOT do:
//
//   - Membership / authorisation. The app owns those records
//     (SPEC §9.6); helpers only ferry events to known device pubkeys.

import { encodeFrame } from "./codec.js";
import {
  AppKeypair,
  X25519_KEY_LEN,
  buildNonce,
  EncryptionError,
} from "./encryption.js";
import {
  buildSigningInput,
  ED25519_SEED_LEN,
  HEADER_DEVICE,
  HEADER_SIG,
  HEADER_TIMESTAMP,
  publicKeyFromSeed,
} from "./signing.js";
import {
  DEFAULT_PUSH_RELAY_URL,
  HEADER_APP,
  PushError,
  PushRejected,
  PushUnavailable,
  type PushConfig,
} from "./push.js";
import * as nodeCrypto from "node:crypto";

export const GROUP_PUSH_PATH = "/group_push";
export const GROUP_PUSH_CONTENT_TYPE = "application/json";

/** Recipient outcomes — must match `sdk/python/pageros/groups.py`. */
export const GROUP_RESULT_ACCEPTED = "accepted";
export const GROUP_RESULT_RATE_LIMITED = "rate_limited";
export const GROUP_RESULT_PAYLOAD_EMPTY = "payload_empty";
export const GROUP_RESULT_PAYLOAD_LARGE = "payload_too_large";
export const GROUP_RESULT_BAD_PAYLOAD = "invalid_payload";
export const GROUP_RESULT_BAD_DEVICE = "invalid_device_pubkey";
export const GROUP_RESULT_STORAGE_ERROR = "storage_error";

export type GroupRecipientResultCode =
  | typeof GROUP_RESULT_ACCEPTED
  | typeof GROUP_RESULT_RATE_LIMITED
  | typeof GROUP_RESULT_PAYLOAD_EMPTY
  | typeof GROUP_RESULT_PAYLOAD_LARGE
  | typeof GROUP_RESULT_BAD_PAYLOAD
  | typeof GROUP_RESULT_BAD_DEVICE
  | typeof GROUP_RESULT_STORAGE_ERROR
  | string; // forward-compat for future codes

/** Per-recipient result returned by the relay. */
export interface GroupRecipientResult {
  device: string; // base64url device pubkey
  result: GroupRecipientResultCode;
  reason: string | undefined;
}

/** Aggregate result of a batch send. */
export interface GroupBroadcastResult {
  status: number;
  accepted: number;
  rejected: number;
  results: GroupRecipientResult[];
}

/** Domain error specific to group broadcast failures. */
export class GroupBroadcastError extends PushError {
  readonly status: number | undefined;
  constructor(message: string, status?: number) {
    super(message);
    this.name = "GroupBroadcastError";
    this.status = status;
  }
}

export interface SendGroupPushOptions {
  /** Wall-clock function. Default `Date.now`. */
  now?: () => number;
  /** Request timeout in ms. Default 10_000. */
  timeoutMs?: number;
  /** `fetch` override for tests. */
  fetchImpl?: typeof fetch;
}

/**
 * Encrypt `payload` once per recipient, sign the batch, and POST to
 * `/group_push` on the Push Relay. Returns the per-recipient verdicts
 * from the relay; throws on transport / signature failures.
 *
 * `recipients` is a list of `[deviceB64, devicePubkeyBytes]` tuples — the
 * b64 form is sent on the wire, the raw bytes feed the per-recipient
 * AEAD. Callers that only have the bytes can derive the b64 via
 * `Buffer.from(pubkey).toString("base64url")`.
 */
export async function sendGroupPush(
  config: PushConfig,
  recipients: Array<[string, Uint8Array]>,
  payload: unknown,
  options: SendGroupPushOptions = {},
): Promise<GroupBroadcastResult> {
  if (config.signingSeed.length !== ED25519_SEED_LEN) {
    throw new GroupBroadcastError(
      `signingSeed must be ${ED25519_SEED_LEN} bytes`,
    );
  }
  publicKeyFromSeed(config.signingSeed);

  const body = buildGroupPushBody(config.keypair, recipients, payload);
  const path = GROUP_PUSH_PATH;
  const base = (config.relayUrl ?? DEFAULT_PUSH_RELAY_URL).replace(/\/+$/, "");
  const url = `${base}${path}`;

  const nowFn = options.now ?? Date.now;
  const ts = String(Math.floor(nowFn() / 1000));
  const signingInput = buildSigningInput("POST", path, ts, body);
  const sig = nodeCrypto.sign(null, signingInput, ed25519PrivateKeyForSeed(config.signingSeed));

  const headers: Record<string, string> = {
    [HEADER_APP]: config.appId,
    [HEADER_DEVICE]: Buffer.from(publicKeyFromSeed(config.signingSeed)).toString("base64url"),
    [HEADER_SIG]: Buffer.from(sig).toString("base64url"),
    [HEADER_TIMESTAMP]: ts,
    "Content-Type": GROUP_PUSH_CONTENT_TYPE,
  };

  const fetchImpl = options.fetchImpl ?? globalThis.fetch;
  if (typeof fetchImpl !== "function") {
    throw new GroupBroadcastError(
      "global fetch is unavailable; pass options.fetchImpl",
    );
  }

  const ac = new AbortController();
  const timer = setTimeout(() => ac.abort(), options.timeoutMs ?? 10_000);
  let response: Response;
  try {
    response = await fetchImpl(url, { method: "POST", headers, body, signal: ac.signal });
  } catch (err) {
    throw new PushUnavailable(
      `push relay unreachable: ${(err as Error).message ?? String(err)}`,
    );
  } finally {
    clearTimeout(timer);
  }

  const responseText = await response.text();
  if (response.status < 200 || response.status >= 300) {
    throw new PushRejected(response.status, responseText);
  }

  // The relay responds with `{accepted, rejected, results: [...]}`
  // (mirrors Python helper). Anything we can't parse is a hard fail.
  let parsed: { accepted?: number; rejected?: number; results?: unknown };
  try {
    parsed = JSON.parse(responseText);
  } catch (err) {
    throw new GroupBroadcastError(
      `push relay returned non-JSON body: ${(err as Error).message}`,
      response.status,
    );
  }

  const results: GroupRecipientResult[] = [];
  if (Array.isArray(parsed.results)) {
    for (const entry of parsed.results) {
      if (entry && typeof entry === "object") {
        const e = entry as Record<string, unknown>;
        results.push({
          device: String(e.device ?? ""),
          result: String(e.result ?? ""),
          reason: e.reason !== undefined ? String(e.reason) : undefined,
        });
      }
    }
  }
  return {
    status: response.status,
    accepted: Number(parsed.accepted ?? results.filter((r) => r.result === GROUP_RESULT_ACCEPTED).length),
    rejected: Number(parsed.rejected ?? results.filter((r) => r.result !== GROUP_RESULT_ACCEPTED).length),
    results,
  };
}

/**
 * Build the JSON body for a `/group_push` batch. Each recipient gets
 * an independent AEAD encryption of the same payload with a unique
 * nonce so devices that can't see each other's traffic still get a
 * fresh ciphertext.
 */
export function buildGroupPushBody(
  keypair: AppKeypair,
  recipients: Array<[string, Uint8Array]>,
  payload: unknown,
): Uint8Array {
  if (recipients.length === 0) {
    throw new GroupBroadcastError("recipients must be non-empty");
  }
  const plaintext = encodePayload(payload);
  const items: Record<string, string>[] = [];
  let counter = 1;
  for (const [b64, pubkey] of recipients) {
    if (pubkey.length !== X25519_KEY_LEN) {
      throw new GroupBroadcastError(
        `recipient ${b64}: pubkey must be ${X25519_KEY_LEN} bytes, got ${pubkey.length}`,
      );
    }
    const nonce = buildNonce(0x02, counter);
    counter += 1;
    let ct: Uint8Array;
    try {
      ct = keypair.encryptTo(pubkey, plaintext, nonce);
    } catch (err) {
      if (err instanceof EncryptionError) {
        throw new GroupBroadcastError(
          `recipient ${b64}: encrypt failed: ${err.message}`,
        );
      }
      throw err;
    }
    items.push({
      device: b64,
      nonce: Buffer.from(nonce).toString("base64url"),
      ct: Buffer.from(ct).toString("base64url"),
    });
  }
  const body = {
    from: Buffer.from(keypair.publicKey).toString("base64url"),
    recipients: items,
  };
  return new TextEncoder().encode(JSON.stringify(body));
}

function encodePayload(payload: unknown): Uint8Array {
  if (payload instanceof Uint8Array) return payload;
  if (typeof payload === "string") return new TextEncoder().encode(payload);
  return encodeFrame(payload);
}

function ed25519PrivateKeyForSeed(seed: Uint8Array): nodeCrypto.KeyObject {
  const header = Buffer.from("302e020100300506032b657004220420", "hex");
  const der = Buffer.concat([header, Buffer.from(seed)]);
  return nodeCrypto.createPrivateKey({ key: der, format: "der", type: "pkcs8" });
}

File: sdk/python/pageros/groups.py
"""Group helpers (PY-007): broadcast + inbound group-event registry.

This is the app-side counterpart to the Push Relay's group fan-out
handler in ``push-relay/internal/server/group.go``. The relay accepts
a single signed envelope and stores N per-recipient encrypted payloads
on the device queues with ``kind: "group_event"`` (SPEC §5.4.2 /
§6.6.6). The relay sees only opaque per-recipient blobs — the same
X25519+ChaCha20-Poly1305 end-to-end pattern as :mod:`pageros.push`
(SPEC §6.6.3, crypto-suite.md §1.1–§1.2).

This module owns the wire format for ``POST /group_push``:

.. code-block:: json

    {
      "recipients": [
        {"device_pubkey": "<b64url>", "payload_b64": "<b64url>"},
        ...
      ]
    }

Each ``payload_b64`` is the same opaque envelope produced by
:func:`pageros.push.build_push_body` — but bound to *that* recipient's
X25519 pubkey, with a per-(app, device) monotonic counter. The
``payload_b64`` field uses raw (un-padded) URL-safe base64 to match the
relay's canonical decoder.

Per-recipient results are returned verbatim from the relay so callers
can distinguish accepted enqueues from rate-limit denials or oversized
payloads on a per-device basis without re-deriving the relay's status
strings.
"""

from __future__ import annotations

import base64
import json
import socket
import time
import urllib.error
import urllib.request
from dataclasses import dataclass, field
from typing import Any, Callable, Iterable, Sequence

from pageros.push import (
    HEADER_APP,
    PushConfig,
    PushError,
    PushRejected,
    PushUnavailable,
    _b64url,
    _coerce_device_pubkey,
)
from pageros.signing import (
    HEADER_SIG,
    HEADER_TIMESTAMP,
    build_signing_input,
)

__all__ = [
    "GROUP_PUSH_PATH",
    "GROUP_PUSH_CONTENT_TYPE",
    "GROUP_RESULT_ACCEPTED",
    "GROUP_RESULT_BAD_DEVICE",
    "GROUP_RESULT_BAD_PAYLOAD",
    "GROUP_RESULT_PAYLOAD_EMPTY",
    "GROUP_RESULT_PAYLOAD_LARGE",
    "GROUP_RESULT_RATE_LIMITED",
    "GROUP_RESULT_STORAGE_ERROR",
    "GroupBroadcastError",
    "GroupBroadcastResult",
    "GroupRecipientResult",
    "build_group_push_body",
    "send_group_push",
]

GROUP_PUSH_PATH = "/group_push"
GROUP_PUSH_CONTENT_TYPE = "application/json"

_DEFAULT_TIMEOUT = 10.0

# Per-recipient status strings — pinned to the Go relay (group.go).
GROUP_RESULT_ACCEPTED = "accepted"
GROUP_RESULT_RATE_LIMITED = "rate_limited"
GROUP_RESULT_PAYLOAD_EMPTY = "payload_empty"
GROUP_RESULT_PAYLOAD_LARGE = "payload_too_large"
GROUP_RESULT_BAD_PAYLOAD = "invalid_payload"
GROUP_RESULT_BAD_DEVICE = "invalid_device_pubkey"
GROUP_RESULT_STORAGE_ERROR = "storage_error"


class GroupBroadcastError(PushError):
    """Raised when a broadcast cannot even be attempted.

    Distinct from :class:`pageros.push.PushRejected` / ``PushUnavailable``
    which still apply when the relay rejects the *whole* batch (auth,
    sig, banned app, body too big). This class is for client-side
    misconfiguration: no recipients, missing app id/key, …
    """


@dataclass(frozen=True)
class GroupRecipientResult:
    """One row of the per-recipient relay response.

    Mirrors the JSON shape returned by ``POST /group_push``. Either
    ``id`` + ``enqueued_at`` are populated (status == ``"accepted"``)
    or ``retry_after`` is populated (status == ``"rate_limited"``);
    other failure statuses leave both unset.
    """

    device_pubkey: str
    status: str
    id: str | None = None
    enqueued_at: int | None = None
    retry_after: int | None = None

    @property
    def accepted(self) -> bool:
        return self.status == GROUP_RESULT_ACCEPTED


@dataclass(frozen=True)
class GroupBroadcastResult:
    """Aggregate result for a single ``broadcast`` call."""

    results: tuple[GroupRecipientResult, ...] = field(default_factory=tuple)

    @property
    def accepted(self) -> tuple[GroupRecipientResult, ...]:
        return tuple(r for r in self.results if r.accepted)

    @property
    def rejected(self) -> tuple[GroupRecipientResult, ...]:
        return tuple(r for r in self.results if not r.accepted)

    @property
    def total(self) -> int:
        return len(self.results)

    def by_device(self, device_pubkey: str) -> GroupRecipientResult | None:
        for r in self.results:
            if r.device_pubkey == device_pubkey:
                return r
        return None


# --------------------------------------------------------------------------- #
# Body framing
# --------------------------------------------------------------------------- #


def build_group_push_body(
    recipients: Sequence[tuple[str, bytes]],
) -> bytes:
    """Build the JSON envelope for ``POST /group_push``.

    ``recipients`` is a sequence of ``(device_pubkey_b64url, opaque_payload_bytes)``
    pairs. Each opaque payload should be the output of
    :func:`pageros.push.build_push_body` for that recipient — the relay
    treats the bytes as a black box and never inspects them.
    """
    if not recipients:
        raise GroupBroadcastError("broadcast requires at least one recipient")
    body = {
        "recipients": [
            {
                "device_pubkey": dev_b64,
                "payload_b64": _b64url(payload),
            }
            for dev_b64, payload in recipients
        ]
    }
    return json.dumps(body, separators=(",", ":")).encode("utf-8")


# --------------------------------------------------------------------------- #
# HTTP
# --------------------------------------------------------------------------- #


def send_group_push(
    config: PushConfig,
    recipients: Sequence[tuple[str, bytes]],
    *,
    timeout: float | None = _DEFAULT_TIMEOUT,
    now: Callable[[], float] | None = None,
    opener: Callable[..., Any] | None = None,
) -> GroupBroadcastResult:
    """Sign and POST a group-event batch to the Push Relay.

    Auth + signature failures (4xx) raise :class:`PushRejected` — those
    failures reject the *whole* batch, matching the relay's behaviour.
    Per-recipient failures (rate limit, bad device key, oversized
    payload) are reported in the returned
    :class:`GroupBroadcastResult` and do not raise.
    """
    body = build_group_push_body(recipients)
    url = config.relay_url.rstrip("/") + GROUP_PUSH_PATH

    clock = now if now is not None else time.time
    ts = str(int(clock()))
    sig_input = build_signing_input("POST", GROUP_PUSH_PATH, ts, body)
    sig = config.signing_key.sign(sig_input).signature

    headers = {
        HEADER_APP: config.app_id,
        HEADER_SIG: _b64url(sig),
        HEADER_TIMESTAMP: ts,
        "Content-Type": GROUP_PUSH_CONTENT_TYPE,
        "Content-Length": str(len(body)),
    }
    req = urllib.request.Request(url, data=body, method="POST", headers=headers)

    fetch = opener if opener is not None else urllib.request.urlopen
    try:
        with fetch(req, timeout=timeout) as resp:
            status = getattr(resp, "status", None) or resp.getcode()
            resp_body = resp.read()
    except urllib.error.HTTPError as exc:
        status = exc.code
        try:
            resp_body = exc.read()
        except Exception:  # pragma: no cover - paranoid
            resp_body = b""
        if 400 <= status < 500:
            raise PushRejected(status, resp_body) from exc
        raise PushUnavailable(
            f"push relay returned HTTP {status}",
            status=status,
            cause=exc,
        ) from exc
    except (urllib.error.URLError, socket.error, OSError) as exc:
        raise PushUnavailable(
            f"push relay unreachable: {exc}", cause=exc
        ) from exc

    if status != 202:
        raise PushUnavailable(
            f"push relay returned unexpected status {status}",
            status=status,
        )

    try:
        parsed = json.loads(resp_body or b"{}")
    except ValueError as exc:
        raise PushUnavailable(
            f"push relay returned malformed JSON: {exc}", cause=exc
        ) from exc

    if not isinstance(parsed, dict):
        raise PushUnavailable("group push response is not a JSON object")
    raw_results = parsed.get("results")
    if not isinstance(raw_results, list):
        raise PushUnavailable("group push response missing 'results' array")

    results: list[GroupRecipientResult] = []
    for row in raw_results:
        if not isinstance(row, dict):
            raise PushUnavailable("group push 'results' row is not a JSON object")
        dev = row.get("device_pubkey")
        st = row.get("status")
        if not isinstance(dev, str) or not isinstance(st, str):
            raise PushUnavailable(
                "group push row missing 'device_pubkey' or 'status'"
            )
        nid = row.get("id")
        enq = row.get("enqueued_at")
        retry = row.get("retry_after")
        results.append(
            GroupRecipientResult(
                device_pubkey=dev,
                status=st,
                id=nid if isinstance(nid, str) and nid else None,
                enqueued_at=int(enq) if isinstance(enq, (int, float)) else None,
                retry_after=int(retry) if isinstance(retry, (int, float)) else None,
            )
        )
    return GroupBroadcastResult(results=tuple(results))


# --------------------------------------------------------------------------- #
# Member-key normalisation
# --------------------------------------------------------------------------- #


def normalize_device_pubkey(device_id: bytes | str) -> tuple[bytes, str]:
    """Return ``(raw_32B, b64url_path_form)`` for any of the accepted forms.

    Thin wrapper over :func:`pageros.push._coerce_device_pubkey` so the
    group registry and the broadcast helper share one validator without
    poking at a private symbol from another module.
    """
    return _coerce_device_pubkey(device_id)
